keps/sig-api-machinery/20190426-crd-defaulting.md

```diff
@@ -0,0 +1,298 @@
+* Mutating Admission Webhooks can be used for defaulting, but:
+  * they are not adequat as it is not possible to set default values of newly added fields on GET because admission is not run in the storage layer when reading from etcd.
+  * webhooks are complex, both from the development/maintenance point of view and due to their non-trivial operational overhead. This makes them a no-go for many "not so ambitiously, professionally developed CRDs", e.g. in in-house enterprise environments.
+* _Structural schemas_ as defined in [KEP Vanilla OpenAPI Subset: Structural Schema](https://github.com/kubernetes/enhancements/pull/1002) makes defaulting a low hanging fruit.
+ 
+### Goals
+
+* add CustomResource defaulting at the correct position in the request pipeline 
+* allow definition of defaults via the OpenAPI v3 `default` field.
+
+### Non-Goals
+
+* allow non-constant defaults: native Golang code can of course set defaults which depends on other fields of a JSON object. This is out of scope here and would need some kind of defaulting webhook.
+* native-type declarative defaulting: this KEP is about CRDs. Though it might be desirable to support the same `// +default=<some-json-value>` tag and a mechanism in `k8s.io/apiserver` to evaluate defaults in native, generic registries, this is out-of-scope of the KEP though.
+
+## Proposal
+
+We assume the CRD has _structural schemas_ (as defined in [KEP Vanilla OpenAPI Subset: Structural Schema](https://github.com/kubernetes/enhancements/pull/1002)).
+
+We propose to
+
+1. derive the value-validation-less variant of the structural schema (trivial by definition of _structural schema_) and 
+2. recursively follow the given CustomResource instance and the structural schema, applying defaults where an object field is 
+  * undefined (`_, ok := obj[field]; !ok`)
+  * `nil` if the field not nullable.
+
+We do this in the serializer by passing a real defaulter to [`versioningserializer.NewCodec`](https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/serializer/versioning/versioning.go#L49) such that defaulting is done natively just after the binary payload has been unmarshalled into an `map[string]interface{}` and pruning of [KEP: Pruning for CustomResources](https://github.com/kubernetes/enhancements/pull/709) was done, compare the yellow boxes in the following figure:
+
+![Decoding steps which must apply defaults](20190426-crd-defaulting-pipeline.png)
+
+We rely on the validation steps in the request pipeline to verify that the default value is of the right type.
+
+The `default` field in the CRD types is considered alpha quality. As the behaviour for CRDs without default values does not change, we will enable defaulting without a feature gate.
+
+[Kubebuilder's crd-gen](https://github.com/kubernetes-sigs/controller-tools/tree/master/cmd/crd) can make use of this feature by adding another tag, e.g. `// +default=<some-json-value>`.
```

Comment:
you should clarify how much structure is allowed.  The examples seem to imply that full arbitrary JSON can be used, but it'd be good to say that.

Reply:
fixed